### ISSUE TITLE
Auto-Masking Stationary Pixels 

### ISSUE THREAD

**Issue by xuwanpeng12** [nianticlabs/monodepth2#264]
Dear:
        First of all, thank you for your amazing paper Monodepth2.
        I encountered a problem when I read your public code.
        As mentioned in the article, you employ Auto-Mask to process stationary pixels. But I did not find the application of the Auto-Mask in your code.

        combined = torch.cat((identity_reprojection_loss, reprojection_loss), dim=1)
        to_optimise, idxs = torch.min(combined, dim=1)
        loss += to_optimise.mean()
     
 
        If  following the above statement, the function of the template is not reflected in the loss. 
        Only the minimum value of reprojection_losses is involved in the calculation.
        Is there something wrong with my understanding, or I am not finding the right place?
         Looking forward to your response, thank you very much！

**Comment by mdfirman** [comment]
Hi there,

Thanks for your interest! Automasking is there, but I implemented it implicitly rather than explicitly. 
Because the min is being taken over reprojection loss _and_ identity reprojection loss, that line is simultaneously computing min reprojection and automasking.

We could have implemented it explicitly, similar to:
```
min_reprojection_loss = torch.min(reprojection_loss, dim=1)
min_identity_reprojection_loss = torch.min(identity_reprojection_loss, dim=1)
automask = (min_reprojection_loss < min_identity_reprojection_loss).float()
to_optimise = min_reprojection_loss * automask
loss += to_optimise.mean() 
```
Which would have done the same thing but with more compute.

See this comment for more details:
https://github.com/nianticlabs/monodepth2/issues/22#issuecomment-503782974

Hope this helps!

**Comment by xuwanpeng12** [comment]
I have learned it. Thanks for your kind words.
